iscsi/targetd/cmd/start.go

```diff
@@ -0,0 +1,135 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	"github.com/kubernetes-incubator/external-storage/iscsi/targetd/provisioner"
+	"github.com/kubernetes-incubator/external-storage/lib/controller"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"time"
+)
+
+var log = logrus.New()
+
+// start-controllerCmd represents the start-controller command
+var startcontrollerCmd = &cobra.Command{
+	Use:   "start",
+	Short: "A brief description of your command",
+	Long: `A longer description that spans multiple lines and likely contains examples
+and usage of using your command. For example:
+
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		initLog()
+
+		log.Debugln("start called")
+
+		// creates the in-cluster config
+		log.Debugln("creating in cluster default kube client config")
+		config, err := rest.InClusterConfig()
+		if err != nil {
+			log.Fatalln(err)
+		}
+		log.WithFields(logrus.Fields{
+			"config-host": config.Host,
+		}).Debugln("kube client config created")
+
+		//	 creates the clientset
+		log.Debugln("creating kube client set")
+		kubernetesClientSet, err := kubernetes.NewForConfig(config)
+		if err != nil {
+			log.Fatalln(err)
+		}
+		log.Debugln("kube client set created")
+
+		// The controller needs to know what the server version is because out-of-tree
+		// provisioners aren't officially supported until 1.5
+		serverVersion, err := kubernetesClientSet.Discovery().ServerVersion()
+		if err != nil {
+			log.Fatalf("Error getting server version: %v", err)
+		}
+
+		//url := viper.GetString("targetd-scheme") + "://" + viper.GetString("targetd-username") + ":" + viper.GetString("targetd-password") + "@" + viper.GetString("targetd-address") + ":" + viper.GetInt("targetd-port")
+
+		url := fmt.Sprintf("%s://%s:%s@%s:%d/targetrpc", viper.GetString("targetd-scheme"), viper.GetString("targetd-username"), viper.GetString("targetd-password"), viper.GetString("targetd-address"), viper.GetInt("targetd-port"))
+
+		log.Debugln("targed URL", url)
+
+		iscsiProvisioner := provisioner.NewiscsiProvisioner(url)
+		log.Debugln("iscsi provisioner created")
+		pc := controller.NewProvisionController(kubernetesClientSet, viper.GetDuration("resync-period"), viper.GetString("provisioner-name"), iscsiProvisioner, serverVersion.GitVersion,
```

Comment:
disclaimer: i am no git expert either.

The diverged state is normal because 'rebase' means, basically, move your commits on top of all the other new commits in upstream/master, effectively changing history since originally your commits came first. So you will have to make sure everything looks okay, maybe backup origin/iscsi-targetd2 into another branch just in case, and then append '--force-with-lease' to 'git push' in order to change history.

When I say the build doesn't work for me from a clean GOPATH, I mean if you clone your code into say /tmp/src/... and you set GOPATH=/tmp does building it still work? It doesn't for me, possibly because my GOPATH does not have the same version packages as you; instead we should be relying entirely on "vendor". So we need to make sure I/anyone cloning the repo can reliably reproduce the build with only the "vendor" directory. We check in the entire vendor directory and don't instruct people to use glide to rebuild it because a) it can take a long time and b) glide is not ubiquitous and we can't be sure cloners know how to use it.